### ISSUE TITLE
Stuck/deadlocked workers in Celery 3.1.x with Redis broker

### ISSUE THREAD

**Issue by mikesun** [celery/celery#2464]
Celery workers become stuck/deadlocked when using Redis broker in Celery 3.1.17.
Reproduced in 3.1.0 and 3.1.16 as well.

**Issue does not occur in current Celery master (3.2.0a2).**
**Issue does not occur with RabbitMQ as broker.**
**Reproducable on Linux and Mac OS.**

### Environment
* Scientific Linux release 6.5 or Mac OS 10.10.2
* Python 2.7.9
* Redis 2.8.9
* Celery 3.1.0, 3.1.16, 3.1.17
* virtualenv 12.0.5

```
software -> celery:3.1.17 (Cipater) kombu:3.0.24 py:2.7.9
            billiard:3.3.0.19 redis:2.10.3
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:disabled

BROKER_URL: 'redis://localhost:6380/0'
```

**redis.conf**
```
# close the connection after a client is idle for N seconds
timeout 0
````

**celeryconfig.py**
```
BROKER_URL = 'redis://localhost:6380/0'
```

**tasks.py**
```
from celery import Celery
import celeryconfig

import time

app = Celery('tasks')
app.config_from_object(celeryconfig)

@app.task
def myTask(id):
    return id
```

**test_client.py**
```
from multiprocessing import Pool
from tasks import myTask 

def run(task_id):
   myTask.delay(1)

num_tasks = 20000
pool = Pool(processes=4)
pool.map(run, xrange(num_tasks))
pool.close()
pool.join()
```

### Reproducing
1. Ran test_client.py to submit 20K tasks.

2. Started two concurrent workers (`-Ofair` doesn't make a difference):
```
celery -A tasks worker --loglevel=DEBUG --concurrency=3 [-Ofair]
```

3. After ~10-20K of processed tasks, workers stuck/deadlocked.
```
[2015-01-21 17:40:31,304: DEBUG/MainProcess] Task accepted: tasks.regulatorTask[6d2be28f-4d97-43b3-84dd-ad28b3a9d0f7] pid:13242
[2015-01-21 17:40:31,305: DEBUG/MainProcess] Task accepted: tasks.regulatorTask[6d798559-3e1e-4fe8-a813-065696d8a1df] pid:13243
[2015-01-21 17:40:31,306: INFO/MainProcess] Received task: tasks.regulatorTask[6adbe350-2d5e-4a92-89d9-4bbd63973aa6]
[2015-01-21 17:40:31,306: DEBUG/MainProcess] TaskPool: Apply <function _fast_trace_task at 0x7f3436baf848> (args:('tasks.regulatorTask', '6adbe350-2d5e-4a92-89d9-4bbd63973aa6', (1,), {}, {'utc': True, u'is_eager': False, 'chord': None, u'group': None, 'args': (1,), 'retries': 0, u'delivery_info': {u'priority': 0, u'redelivered': None, u'routing_key': u'celery', u'exchange': u'celery'}, 'expires': None, u'hostname': 'celery@dev6-cig-c53d391f.ewr01.tumblr.net', 'task': 'tasks.regulatorTask', 'callbacks': None, u'correlation_id': u'6adbe350-2d5e-4a92-89d9-4bbd63973aa6', 'errbacks': None, 'timelimit': (None, None), 'taskset': None, 'kwargs': {}, 'eta': None, u'reply_to': u'7d9537ab-0a15-34eb-afa2-aafc60e63b1a', 'id': '6adbe350-2d5e-4a92-89d9-4bbd63973aa6', u'headers': {}}) kwargs:{})
[2015-01-21 17:40:31,307: INFO/MainProcess] Task tasks.regulatorTask[6d2be28f-4d97-43b3-84dd-ad28b3a9d0f7] succeeded in 0.00572886499867s: 1
[2015-01-21 17:40:31,308: INFO/MainProcess] Task tasks.regulatorTask[6d798559-3e1e-4fe8-a813-065696d8a1df] succeeded in 0.00644170500163s: 1
```

`strace` indicates workers blocking on I/O read from Redis
```
Process 13241 attached - interrupt to quit
read(13,

Process 13242 attached - interrupt to quit
read(17,

Process 13243 attached - interrupt to quit
read(9,
```

### Notables
* adding a connection `timeout` in Redis will cause deadlock to release after timeout period
* specifying a `CELERYD_TASK_TIME_LIMIT` does not prevent/release the deadlock
* single worker execution (`--concurrency=1`) works properly

**Comment by joostdevries** [comment]
I may be facing a similar issue with 3.1.18. Worker stops processing new tasks without any kind of warning. Turned on debug logging and found out that new tasks were still being accepted. Redis as broker on Ubuntu 12

**Comment by raghuramos1987** [comment]
can you post your client-output-buffer-limit option value on the redis server?

**Comment by raghuramos1987** [comment]
i might be facing a similar issue. unfortunately, i'm unable to reproduce it in dev. but, when i use 3.1.0 instead of 3.1.18 things seem to work better (till now).

**Comment by raghuramos1987** [comment]
according to https://pypi.python.org/pypi/redis/ ...

"Note that, because the Pipeline must bind to a single connection for the duration of a WATCH, care must be taken to ensure that the connection is returned to the connection pool by calling the reset() method. If the Pipeline is used as a context manager (as in the example above) reset() will be called automatically. Of course you can do this the manual way by explicity calling reset():"

And, in the code, we dont call reset (AFAIK). I'll test with that change and see if it fixes it.

**Comment by raghuramos1987** [comment]
can you see if this fixes it? (it hasnt failed yet for me. but, i'd like to know how it can be stress tested).
```
diff --git a/celery/backends/redis.py b/celery/backends/redis.py
index 236ac38..f151bbd 100644
--- a/celery/backends/redis.py
+++ b/celery/backends/redis.py
@@ -164,8 +164,8 @@ class RedisBackend(KeyValueStoreBackend):
             pipe.setex(key, value, self.expires)
         else:
             pipe.set(key, value)
-        pipe.publish(key, value)
         pipe.execute()
+        pipe.reset()

     def delete(self, key):
         self.client.delete(key)
```

**Comment by thedrow** [comment]
@raghuramos1987 Care to provide a pull request with this change?

**Comment by raghuramos1987** [comment]
actually. i looked more into this. pipe.execute should reset it anyways. so, it might not fix the issue. still looking.

**Comment by raghuramos1987** [comment]
@thedrow - i've a pull request with a more probable solution at https://github.com/celery/celery/pull/2679

**Comment by trevor-h** [comment]
I too experience this issue with a Beat enabled worker with jobs running every few seconds. It works for days/weeks then all of a sudden stops processing jobs and creating a backlog on the Redis broker. Simply restarting the worker resolves the issue.

It's worth noting that I have several environments running an application and only 1 of these 5 environments has experienced the issue (repeatedly).

The single worker with several issues just this month:
python 2.6.6
billiard==3.3.0.19
celery==3.1.17
kombu==3.0.24
redis==2.10.3

The workers with no issues (yet) after many months:
python 2.6.6
billiard==3.3.0.20
celery==3.1.18
kombu==3.0.26
redis==2.10.3

I haven't looked into the differences in these versions deeper than that, though it is interesting to me I've seen 3 issues in 4 weeks on a single server (with the older config) and never seen it on other systems once.

**Comment by raghuramos1987** [comment]
@trevor-h please look at https://github.com/celery/kombu/pull/490 . you need to set a socket timeout if the symptoms can be verified as stated there.

**Comment by yatharthsharma** [comment]
Hey ,

I am still facing the issue. setting socket_timeout didn't help.

celery report 

software -> celery:3.1.18 (Cipater) kombu:3.0.26 py:2.7.6
            billiard:3.3.0.20 py-amqp:1.4.7
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.default.Loader
settings -> transport:amqp results:disabled


broker - redis
app = Celery('tasks', broker = 'redis://localhost:6379/0')

It works fine for around 1 days and then it halts. Can you please help me debug it?

**Comment by yatharthsharma** [comment]
Hi,

Further Information: .Mostly the jobs are calling twitter/fb API

celery report ->

software -> celery:3.1.18 (Cipater) kombu:3.0.26 py:2.7.6
billiard:3.3.0.20 py-amqp:1.4.7
platform -> system:Linux arch:64bit, ELF imp:CPython
loader -> celery.loaders.default.Loader
settings -> transport:amqp results:disabled

Here is my config ->

{
"ADMINS": [], 
"BROKER_CONNECTION_MAX_RETRIES": null, 
"BROKER_CONNECTION_RETRY": true, 
"BROKER_CONNECTION_TIMEOUT": 4, 
"BROKER_FAILOVER_STRATEGY": null, 
"BROKER_HEARTBEAT": null, 
"BROKER_HEARTBEAT_CHECKRATE": 3.0, 
"BROKER_HOST": null, 
"BROKER_LOGIN_METHOD": null, 
"BROKER_PASSWORD": "*****", 
"BROKER_POOL_LIMIT": null, 
"BROKER_PORT": null, 
"BROKER_TRANSPORT": null, 
"BROKER_TRANSPORT_OPTIONS": {
"SOCKET_TIMEOUT": 3600
}, 
"BROKER_URL": "redis://localhost:6379/0", 
"BROKER_USER": null, 
"BROKER_USE_SSL": false, 
"BROKER_VHOST": null, 
"CASSANDRA_COLUMN_FAMILY": null, 
"CASSANDRA_DETAILED_MODE": false, 
"CASSANDRA_KEYSPACE": "**", 
"CASSANDRA_READ_CONSISTENCY": null, 
"CASSANDRA_SERVERS": null, 
"CASSANDRA_WRITE_CONSISTENCY": null, 
"CELERYBEAT_LOG_FILE": null, 
"CELERYBEAT_LOG_LEVEL": "INFO", 
"CELERYBEAT_MAX_LOOP_INTERVAL": 0, 
"CELERYBEAT_SCHEDULE": {}, 
"CELERYBEAT_SCHEDULER": "celery.beat:PersistentScheduler", 
"CELERYBEAT_SCHEDULE_FILENAME": "celerybeat-schedule", 
"CELERYBEAT_SYNC_EVERY": 0, 
"CELERYD_AGENT": null, 
"CELERYD_AUTORELOADER": "celery.worker.autoreload:Autoreloader", 
"CELERYD_AUTOSCALER": "celery.worker.autoscale:Autoscaler", 
"CELERYD_CONCURRENCY": 0, 
"CELERYD_CONSUMER": "celery.worker.consumer:Consumer", 
"CELERYD_FORCE_EXECV": false, 
"CELERYD_HIJACK_ROOT_LOGGER": true, 
"CELERYD_LOG_COLOR": null, 
"CELERYD_LOG_FILE": null, 
"CELERYD_LOG_FORMAT": "[%(asctime)s: %(levelname)s/%(processName)s] %(message)s", 
"CELERYD_LOG_LEVEL": "WARN", 
"CELERYD_MAX_TASKS_PER_CHILD": null, 
"CELERYD_POOL": "prefork", 
"CELERYD_POOL_PUTLOCKS": true, 
"CELERYD_POOL_RESTARTS": false, 
"CELERYD_PREFETCH_MULTIPLIER": 4, 
"CELERYD_STATE_DB": null, 
"CELERYD_TASK_LOG_FORMAT": "[%(asctime)s: %(levelname)s/%(processName)s] %(task_name)s[%(task_id)s]: %(message)s", 
"CELERYD_TASK_SOFT_TIME_LIMIT": null, 
"CELERYD_TASK_TIME_LIMIT": null, 
"CELERYD_TIMER": null, 
"CELERYD_TIMER_PRECISION": 1.0, 
"CELERYD_WORKER_LOST_WAIT": 10.0, 
"CELERYMON_LOG_FILE": null, 
"CELERYMON_LOG_FORMAT": "[%(asctime)s: %(levelname)s] %(message)s", 
"CELERYMON_LOG_LEVEL": "INFO", 
"CELERY_ACCEPT_CONTENT": [
"json", 
"pickle", 
"msgpack", 
"yaml"
], 
"CELERY_ACKS_LATE": false, 
"CELERY_ALWAYS_EAGER": false, 
"CELERY_ANNOTATIONS": null, 
"CELERY_BROADCAST_EXCHANGE": "celeryctl", 
"CELERY_BROADCAST_EXCHANGE_TYPE": "fanout", 
"CELERY_BROADCAST_QUEUE": "celeryctl", 
"CELERY_CACHE_BACKEND": null, 
"CELERY_CACHE_BACKEND_OPTIONS": {}, 
"CELERY_CHORD_PROPAGATES": true, 
"CELERY_COUCHBASE_BACKEND_SETTINGS": null, 
"CELERY_CREATE_MISSING_QUEUES": true, 
"CELERY_DEFAULT_DELIVERY_MODE": 2, 
"CELERY_DEFAULT_EXCHANGE": "celery", 
"CELERY_DEFAULT_EXCHANGE_TYPE": "direct", 
"CELERY_DEFAULT_QUEUE": "celery", 
"CELERY_DEFAULT_RATE_LIMIT": null, 
"CELERY_DEFAULT_ROUTING_KEY": "**", 
"CELERY_DISABLE_RATE_LIMITS": false, 
"CELERY_EAGER_PROPAGATES_EXCEPTIONS": false, 
"CELERY_ENABLE_REMOTE_CONTROL": true, 
"CELERY_ENABLE_UTC": true, 
"CELERY_EVENT_QUEUE_EXPIRES": null, 
"CELERY_EVENT_QUEUE_TTL": null, 
"CELERY_EVENT_SERIALIZER": "json", 
"CELERY_IGNORE_RESULT": false, 
"CELERY_IMPORTS": [], 
"CELERY_INCLUDE": [
"celery.app.builtins", 
"tasks"
], 
"CELERY_MAX_CACHED_RESULTS": 100, 
"CELERY_MESSAGE_COMPRESSION": null, 
"CELERY_MONGODB_BACKEND_SETTINGS": null, 
"CELERY_QUEUES": null, 
"CELERY_QUEUE_HA_POLICY": null, 
"CELERY_REDIRECT_STDOUTS": true, 
"CELERY_REDIRECT_STDOUTS_LEVEL": "WARNING", 
"CELERY_REDIS_DB": null, 
"CELERY_REDIS_HOST": null, 
"CELERY_REDIS_MAX_CONNECTIONS": null, 
"CELERY_REDIS_PASSWORD": "**", 
"CELERY_REDIS_PORT": null, 
"CELERY_RESULT_BACKEND": null, 
"CELERY_RESULT_DBURI": null, 
"CELERY_RESULT_DB_SHORT_LIVED_SESSIONS": false, 
"CELERY_RESULT_DB_TABLENAMES": null, 
"CELERY_RESULT_ENGINE_OPTIONS": null, 
"CELERY_RESULT_EXCHANGE": "celeryresults", 
"CELERY_RESULT_EXCHANGE_TYPE": "direct", 
"CELERY_RESULT_PERSISTENT": null, 
"CELERY_RESULT_SERIALIZER": "pickle", 
"CELERY_ROUTES": null, 
"CELERY_SECURITY_CERTIFICATE": null, 
"CELERY_SECURITY_CERT_STORE": null, 
"CELERY_SECURITY_KEY": "**", 
"CELERY_SEND_EVENTS": false, 
"CELERY_SEND_TASK_ERROR_EMAILS": false, 
"CELERY_SEND_TASK_SENT_EVENT": false, 
"CELERY_STORE_ERRORS_EVEN_IF_IGNORED": false, 
"CELERY_TASK_PUBLISH_RETRY": true, 
"CELERY_TASK_PUBLISH_RETRY_POLICY": {}, 
"CELERY_TASK_RESULT_EXPIRES": "1 day, 0:00:00", 
"CELERY_TASK_SERIALIZER": "pickle", 
"CELERY_TIMEZONE": null, 
"CELERY_TRACK_STARTED": false, 
"CELERY_WORKER_DIRECT": false, 
"EMAIL_HOST": "localhost", 
"EMAIL_HOST_PASSWORD": "*****", 
"EMAIL_HOST_USER": null, 
"EMAIL_PORT": 25, 
"EMAIL_TIMEOUT": 2, 
"EMAIL_USE_SSL": false, 
"EMAIL_USE_TLS": false, 
"SERVER_EMAIL": "celery@localhost"
}

**Comment by ask** [comment]
Sounds like a duplicate of #2906, as pipe's being introduced greatly increases the chances of connection.disconnect being called.

**Comment by djova** [comment]
We're running a cluster with 10 workers processing about 5000 events per second, using Redis as a broker, and we've been seeing this same deadlock issue fairly consistently for a while now. 

In https://github.com/celery/celery/issues/1770, they mentioned having success resolving the deadlocks using `-Ofair`. We tried it out and the deadlocks completely went away.

**Comment by ask** [comment]
Closing this, as we don't have the resources to complete this task.

May be fixed in master, let's see if comes back after 4.0 release.

**Comment by gzcf** [comment]
It occurs at celery (4.0.2)

**Comment by thedrow** [comment]
@gzcf Can you please open a new issue about it with the details required to reproduce this?

**Comment by nigma** [comment]
I'm also observing this with Celery 4.0.2 which led me to this investigation and looking into stack traces and connection details - https://gist.github.com/nigma/a4e572a3c3aaf50395a704b05b480bbf.

Celery worker pool configured with Redis broker stops responding after processing a fast batch of ~2000 small tasks (out of 4k total).

It goes completely silent but not crashes and it looks it gets stuck at `recvfrom` from Redis connection.

Nodes are running in Docker on AWS ElasticBeanstalk and are connecting to Redis on AWS ElastiCache. There is no sign of any errors in Sentry and New Relic. Memory and CPU usage is within limits.

The settings have set the `ACKS_LATE`, `-Ofair`, prefetch multiplier 1 and `ignore_result` set.

@thedrow I'm happy to open a new issue however there is already a couple related ones and all of them seem to be related to handling Redis broker connections: #2606, #2906, #2904, #3765.
Should we pick one and move the discussion there or do you want me to open a new issue?

**Comment by gzcf** [comment]
@thedrow My case is almost the same as nigma's. For me, Redis broker stops responding after processing a fast batch of small tasks (out of 3k total). Each of tasks will take <50ms. Some may take <5 ms.

Since nigma already had more details than me, it's more suitable to let him to open a new issue.
Also, I am glad to provide my details. When a new issue is opened, just notify me.

**Comment by thedrow** [comment]
@nigma If you're still seeing this issue with Celery 4.x we'd rather see a new issue about it.

**Comment by elcricwu** [comment]
Celery use redis pubsub mechanism. set redis config: 
`client-output-buffer-limit pubsub 512mb 256mb 60 `
to avoid redis server closing the connection.

**Comment by nigma** [comment]
I have narrowed down the problem I'm seeing with 4.0.2 to a conflict with `hiredis` package and opened a new issue in #3898.

@gzcf is this case similar to what you observe?

**Comment by alexsv** [comment]
is it possible to get an official response on this problem (we know about it will fix | we know about it and can't fix now because of resource | we know, but can't reproduce, we can't reproduce). Thanks!

**Comment by gzcf** [comment]
@nigma I don't use `hiredis`

**Comment by ompemi** [comment]
Any updates on this? It makes the system very unstable.

As the initial issue says (redis broker):

- adding a connection timeout in Redis will cause deadlock to release after timeout period
- single worker execution (--concurrency=1) works properly

If we have socket timeout of Redis to 3sec: after processing few thousand tasks the worker gets stuck and it always takes 3seconds or more to process (during an HTTP call that requires a syscall). Running strace and lsof is always waiting on a pipe of type `FIFO` related with Celery.

**Comment by thedrow** [comment]
@alexsv @ompemi We don't have the resources to *debug* it currently. Unless someone provides an analysis that allows us to implement a fix this issue will not be resolved. I don't use the redis broker in production and we currently don't have a maintainer that does.

I'm going to lock this issue because there are too many interleaving problems here, some related to Celery 4.x and some are related to Celery 3.x. Some users are using the gevent pool, others use the eventlet pool and others are using prefork. Each combination of these components may trigger different modes of failures which is confusing us and it does not contribute to solving this issue.
Moreover, for some users using `-Ofair` [resolved their problem](https://github.com/celery/celery/issues/2464#issuecomment-191465323).
This indicates that we're dealing with multiple issues.

If you have an issue with **Celery 4.x** like #3898 please report it using a new issue.

The analysis provided in #3898 indicates that the `worker_max_tasks_per_child` setting triggers this behavior. 
If you encounter a different failure mode, please open a new issue so that we can track fixing these failure modes separately.

**Comment by matt-dalton** [comment]
We have suddenly starting experiencing this with Celery 4.4.7 with Redis. 

It bizarrely just suddenly started happening every few hours. We've tried:
- Removing worker_max_tasks_per_child
- Setting Ofair through celery worker --uid celery -A my_app --loglevel=INFO -E -O fair

Any other recent pearls of wisdom? Did you know if there was ever any progress here @thedrow? Happy to give more details on our setup if that helps.

**Comment by thedrow** [comment]
> 
> 
> We have suddenly starting experiencing this with Celery 4.4.7 with Redis.
> 
> It bizarrely just suddenly started happening every few hours. We've tried:
> 
>     * Removing worker_max_tasks_per_child
> 
>     * Setting Ofair through celery worker --uid celery -A my_app --loglevel=INFO -E -O fair
> 
> 
> Any other recent pearls of wisdom? Did you know if there was ever any progress here @thedrow? Happy to give more details on our setup if that helps.

This issue has not seen any progress since no one has provided an analysis of all the failure modes they encountered.
This issue is actually a number of issues causing the same symptoms as it seems.
I'd like you to try and debug this. If you can come up with a solution, please share it with us.

**Comment by matt-dalton** [comment]
What sort of information would be helpful? The challenging thing is there appears to be no output/error - the logs just stop, so i'm finding it quite difficult to dive into. Are there other sources of information I could look at?

**Comment by thedrow** [comment]
You can try to use the remote debugger to give us a traceback and the situation you are in.
See https://docs.celeryproject.org/en/stable/userguide/debugging.html#enabling-the-break-point-signal.

**Comment by matt-dalton** [comment]
OK cool thanks. Is this safe to run in prod generally speaking? We are only experiencing this situation in our live environment currently.

**Comment by thedrow** [comment]
It's safe as enabling any debugging measure in production, You should disable it as soon as you encounter the problem.

**Comment by matt-dalton** [comment]
To update on this...we haven't yet been able to recreate the issue for the last week, but are gradually increasing prod loads as we gain confidence, so will keep an eye out.
The only fix we did make was to our Redis service - this was being restarted because it was taking up too much memory, so it's possible the workers were not discovering Redis correctly after the restart. However from what I can see in the logs, the Redis restarts don't completely tie in with when we saw the worker issue, so not 100% convinced that's the answer. But possibly a theory to follow if anyone else experiences this.
